src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -510,47 +514,57 @@ private void CheckDisposed()
+
+                        var buffer = result.Buffer;
+                        var consumed = buffer.Start;
+                        var examined = buffer.End;
+
+                        try
                         {
-                            if (HandshakeProtocol.TryParseResponseMessage(ref buffer, out var message))
+                            // Read first message out of the incoming data
+                            if (!buffer.IsEmpty)
                             {
-                                // Adjust consumed and examined to point to the end of the handshake
-                                // response, this handles the case where invocations are sent in the same payload
-                                // as the the negotiate response.
-                                consumed = buffer.Start;
-                                examined = consumed;
-
-                                if (message.Error != null)
+                                if (HandshakeProtocol.TryParseResponseMessage(ref buffer, out var message))
                                 {
-                                    Log.HandshakeServerError(_logger, message.Error);
-                                    throw new HubException(
-                                        $"Unable to complete handshake with the server due to an error: {message.Error}");
-                                }
+                                    // Adjust consumed and examined to point to the end of the handshake
+                                    // response, this handles the case where invocations are sent in the same payload
+                                    // as the the negotiate response.
+                                    consumed = buffer.Start;
+                                    examined = consumed;
+
+                                    if (message.Error != null)
+                                    {
+                                        Log.HandshakeServerError(_logger, message.Error);
+                                        throw new HubException(
+                                            $"Unable to complete handshake with the server due to an error: {message.Error}");
+                                    }
 
-                                break;
+                                    break;
+                                }
+                            }
+                            else if (result.IsCompleted)
+                            {
+                                // Not enough data, and we won't be getting any more data.
+                                throw new InvalidOperationException(
+                                    "The server disconnected before sending a handshake response");
+                            }
+                            else if (result.IsCanceled)
+                            {
+                                throw new OperationCanceledException(
```

Comment:
This isn't necessarily true. Since you're linking 2 tokens, you'll want to catch OperationCancelledException and then check each of the tokens to see which one was cancelled.

Reply:
Also, if we're going to throw an exception because of a timeout elapsing, it should be `TimeoutException`, even if a CancellationToken is how we are triggering it.